src/efftester.ml

```diff
@@ -1089,39 +1089,57 @@ module Generators (Ctx : Context) = struct
     in
     [ (3, gen) ]
 
-  and list_permute_term_gen_inner env goal size rules =
-    let rec remove_at n xs =
-      match (n, xs) with
-      | 0, _ :: xs -> xs
-      | n, x :: xs -> x :: remove_at (n - 1) xs
-      | _ -> failwith "index out of bounds"
-    in
-    let elts_weighted xs =
-      let _, ig =
-        List.fold_left
-          (fun (i, acc) (w, g) -> (i + 1, (w, Gen.pair (Gen.return i) g) :: acc))
-          (0, [])
-          xs
+  (* {!gen_term_from_rules} tries to generate a term of given type (goal) from given rules
+  (rules):
+    1. It sums the weights of all rules
+
+    2. Using own implementation of Gen.frequency called {!gen_freq}, it picks an element
+    from the (weight, rule) list
+
+    3. If the picked rule can generate a term successfully, then the term wrapped in an
+    option type is returned; otherwise, we look for another rule that can do so using
+    {!gen_freq} with the updated list.
+
+    4. If trying to generate a term we run out of rules, we return None.
+  *)
+  and gen_term_from_rules _env _goal _size rules =
+    let weights_sum = List.fold_left (fun acc (w, _g) -> acc + w) 0 rules in
+    (* own implementation of Gen.frequency:
+      returns a randomly chosen element from the given list and an updated list without
+      the chosen element;
+      we need the picked element's rule to try to generate a value and the list to have
+      other rules if this one fails*)
+    let gen_freq lst rand_k =
+      let rand_k = rand_k - 1 in
```

Comment:
I find this definition a bit confusing. Maybe you could *not* decrement `rand_k` here and use a `<=`-test below?